### PR TITLE
feature(unlock-app): adding support for crosschain purchase

### PR DESCRIPTION
**Pull request by julien51**
# Description

Adding cross-chain purchase thru the use of Decent's API to find the routes.

This requires these 2 PRs to be merged/rebased first:
- https://github.com/unlock-protocol/unlock/pull/12787
- https://github.com/unlock-protocol/unlock/pull/12784

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx [hunk 1]
```diff
@@ -32,14 +33,15 @@ export function Confirm({
     ToastHelper.error(message)
   }
 
-  const onConfirmed = (lock: string, hash?: string) => {
+  const onConfirmed = (lock: string, hash?: string, network?: number) => {
```
Comment (julien51) [inline unlock-app/src/components/interface/checkout/main/Confirm.tsx:36]: The conformation screen needs to wait for the transaction to be mined on the "origin" chain, which can be different from the lock's chain so we are passing optionally the network.

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx [hunk 2]
```diff
@@ -73,6 +76,14 @@ export function Confirm({
           onError={onError}
         />
       )}
+      {payment.method === 'crosschain_purchase' && (
```
Comment (julien51) [inline unlock-app/src/components/interface/checkout/main/Confirm.tsx:79]: The new screen!

File: unlock-app/src/components/interface/checkout/main/Minting.tsx [hunk 3]
```diff
@@ -185,16 +190,32 @@ export function Minting({
...
-  const { mint, lock, messageToSign, metadata } = state.context
+  const { mint, lock, messageToSign, metadata, recipients } = state.context
   const processing = mint?.status === 'PROCESSING'
 
   useEffect(() => {
     if (mint?.status !== 'PROCESSING') {
       return
     }
+
+    const waitForTokenIds = async (): Promise<string[]> => {
```
Comment (julien51) [inline unlock-app/src/components/interface/checkout/main/Minting.tsx:201]: In cross chain contexts, waiting for the tx to be mined on the sending chain is not enough. We need to wait for the tokens on the destination chain.

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx (hunk 1 above)
Comment (clemsos) [reply unlock-app/src/components/interface/checkout/main/Confirm.tsx:36]: maybe should be called `transactionNetwork` to be consistent with what is found in the component  ?

File: unlock-app/src/components/interface/checkout/main/Minting.tsx [hunk 4]
```diff
@@ -185,16 +190,32 @@ export function Minting({
...
+      const tokenIds = await Promise.all(
+        recipients.map((r: string) =>
+          web3Service.latestTokenOfOwner(lock!.address, r, lock!.network)
+        )
+      )
+      if (tokenIds.filter((tokenId?: string) => !!tokenId).length) {
+        return tokenIds
+      }
+      await sleeper(1000)
+      return waitForTokenIds()
```
Comment (clemsos) [inline unlock-app/src/components/interface/checkout/main/Minting.tsx:213]: not sure why you are calling that function recursively?

File: unlock-app/src/components/interface/checkout/main/Payment.tsx [hunk 5]
```diff
@@ -91,24 +119,25 @@ export function Payment({ injectedProvider, checkoutService }: Props) {
...
   const enableClaim: boolean = !!(
     canClaim &&
     !isCanClaimLoading &&
     isReceiverAccountOnly &&
-    !balance?.isPayable
+    !balance?.isGasPayable
   )
 
-  const { data: routes, isInitialLoading: isUniswapRoutesLoading } =
+  const canAfford = balance?.isGasPayable && balance?.isPayable
```
Comment (clemsos) [inline unlock-app/src/components/interface/checkout/main/Payment.tsx:138]: will that check if account has enough to pay for gas+price. It seems it checks for it separately

File: unlock-app/src/utils/theBox.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
+import { Lock } from '~/unlockTypes'
+import { ethers } from 'ethers'
+import axios from 'axios'
+import { networks } from '@unlock-protocol/networks'
+
+export interface BoxActionRequest {
```
Comment (clemsos) [inline unlock-app/src/utils/theBox.ts:6]: why is it called the "box"?

**Review by clemsos — APPROVED** [review]

File: unlock-app/src/components/interface/checkout/main/Payment.tsx (hunk 5 above)
Comment (julien51) [reply unlock-app/src/components/interface/checkout/main/Payment.tsx:138]: Indeed, we need to be better eventually to check for the actual gas spent. 
- For non native tokens it checks that the balance is enough + there is _some_ native to pay for gas.
- For native tokens it checks there is indeed more than the cost of the membership, but not that there is enough to _also_ pay for gas if there is just enough to pay the membership.
The wallet will show error messages in both cases though.

File: unlock-app/src/utils/theBox.ts (hunk 6 above)
Comment (julien51) [reply unlock-app/src/utils/theBox.ts:6]: That's Decent's lingo :/

**Pull request merged**